### DISCUSSION TITLE
Nested / complex pattern matching

### DISCUSSION THREAD

**Discussion by nohehf** [gleam-lang/gleam#3958] [Ideas & suggestions]
Hello ! 
I've seen that you can pattern match nested lists (lists in lists):
```gleam
fn get_first_non_empty(lists: List(List(t))) -> List(t) {
  case lists {
    [[_, ..] as first, ..] -> first
    [_, ..rest] -> get_first_non_empty(rest)
    [] -> []
  }
}
```

but it seems that we cannot do it with different types such as:
```gleam
fn(list: List(Result)) {
  case list {
    [Ok(val)] -> todo // match lists with first value being Ok, and unwrap at the same time (val is accessible here)
   _ -> todo
  }
}
```

Is there a way to do this elegantly ? I think this could be really convenient and is done is some functional languages I believe.
I think it has been a bit discussed here: https://github.com/gleam-lang/gleam/discussions/1271 (and I do agree with the `case _` addition too)

**Comment by lpil** [comment]
You can do that! That code has always been valid in Gleam.

**Reply by nohehf** [reply]
Oh I'm sorry, shoud have tested my minimal reproduction. I was on a more complex case and it was probably not the issue then. Thanks :)

**Reply by nohehf** [reply]
Oh okay I missed the as ! I think adding some more examples to the docs / tour on this could be nice, will do a PR
